### ISSUE TITLE
DataGrid and TreeList - Toolbar customization enhancement

### ISSUE THREAD

**Issue by NickMitrokhin** [DevExpress/DevExtreme#19076]
# The Problem

Currently, the only way to customize toolbar in DataGrid/TreeList is to use [onToolbarPreparing](https://js.devexpress.com/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/#onToolbarPreparing) event. You cannot use this method to change settings of buttons once toolbar is created. Direct DOM mutations are not recommended when you use React, Angular, and Vue frameworks.

# The Proposed Solution
We plan to add a new `toolbar` option that allows you to customize toolbar items the same way as the [Toolbar](https://js.devexpress.com/Documentation/ApiReference/UI_Components/dxToolbar/Configuration/items/). Each toolbar item can be specified as an object with the same number of properties as an item of a stand-alone [toolbar](https://js.devexpress.com/Documentation/ApiReference/UI_Components/dxToolbar/Configuration/items/) or a string name from the following  list: 'addRowButton', 'applyFilterButton','columnChooserButton', 'exportButton', 'groupPanel', 'revertButton', 'saveButton', 'searchPanel'. 
Note that the option will allow you to customize only toolbar items, not the entire toolbar. If you want to customize the toolbar, use the [onToolbarPreparing](https://js.devexpress.com/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/#onToolbarPreparing) event.

The following code snippet shows how to specify items:

**jQuery**
```js
$('#container').dxDataGrid({
    ...
    toolbar: {
        items: [
            {
                location: "before",
                widget: "dxButton",
                options: {
                    text: "My button",
                    width: 120,
                    onClick: function(e) {
                        // execute your code when the button is clicked
                    }
                }
            },
            'columnChooserButton'
        ]
    },
    ...   
})
```

**Angular**
```html
<dx-data-grid ...>
    ...
    <dxo-toolbar>
        <dxi-item location="before">
            <dx-button
                text="My button"
                width="120"
                (onClick)="onClickHandler($event)">
            </dx-button>        
        </dxi-item>
        <dxi-item name="columnChooserButton"></dxi-item>
    </dxo-toolbar>
    ...   
</dx-data-grid>
```

**React**
```html
<DataGrid ...>
    ...
    <Toolbar>
        <Item location='before'>
            <Button
              text='My button'
              width='120'
              onClick={this.onClickHandler} />
        </Item>
        <Item name='columnChooserButton' />
    </Toolbar>        
    ...
</DataGrid>
```

**Vue**
```html
<DxDataGrid ...>
    ...
    <DxToolbar>
        <DxItem location="before">
            <DxButton
                text="My button"
                width="120"
                @click="onClickHandler"
            />
        </DxItem>
        <DxItem
            name="columnChooserButton"
        />
    </DxToolbar>        
    ...
</DxDataGrid>
```

If you need a better solution to customize the whole toolbar, you can share your usage scenario in the comments and we will consider them for future customization API enhancements.

# Try It
## Live Sandbox
[jQuery](https://codepen.io/nickmitrokhin/pen/vYZEoLM?editors=0010)
[React](https://codesandbox.io/s/toolbar-customization-devextreme-data-grid-forked-k7umf)
[Vue](https://codesandbox.io/s/toolbar-customization-devextreme-data-grid-forked-sf574)

# We need Your Feedback

## Take a Quick Poll
[Do the new toolbar option meet your needs?](https://forms.gle/aD49rTcyVtFxkP749)

**Comment by jakehockey10** [comment]
Looking forward to this!  Thank you!

**Comment by softboy99** [comment]
+1

**Comment by Lonli-Lokli** [comment]
Just FYI, in React demo
1. `onToolbarPreparing={this.onToolbarPreparing}` exists but handler is not defined
2. App is broken on Collapse All button click

**Comment by NickMitrokhin** [comment]
Thank you for pointing out these issues. We updated the example.

**Comment by reezan** [comment]
is this available on 21.6?

**Comment by NickMitrokhin** [comment]
Hi @reezan,

This feature will be available only in v21.2.

**Comment by dean-shanghai** [comment]
Toolbar customization only for DataGrid, not TreeList?

**Comment by alasoft** [comment]
I have a detected a possible problem with new dxDataGrid toolbar implementation, can I comment that here ?

Robert

**Comment by jakehockey10** [comment]
I'm attempting to use this for the first time, and I can't get the `addRowButton` customization to work:

``` html
            <dxi-item name="addRowButton"
                      [showText]="'always'"
                      [options]="{icon: null, text: 'Create', type: 'success', stylingMode: 'contained'}">
            </dxi-item>
```

Despite what I put in the options, it still has the default look and feel for the material theme.  Am I not using the options correctly?

**Comment by onebalaban** [comment]
@alasoft sure

**Comment by Alyar666** [comment]
These features are now available in the [v21.2](https://github.com/DevExpress/DevExtreme/releases/tag/21.2.3) release. I'm closing this thread. In the case of bugs or questions, feel free to create a new ticket in our [Support Center](https://supportcenter.devexpress.com/).
